### PR TITLE
DYN-899: Make symbols local to imperative block scope

### PR DESCRIPTION
**Pull request by aparajit-pratap**
### Purpose

Variables passed inside imperative blocks from an outer scope will be read only.

JIRA: https://jira.autodesk.com/browse/DYN-899
image

### Declarations

Check these if you believe they are true

- [x] The code base is in a better state after this PR
- [ ] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [x] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [x] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions), and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.

### Reviewers

@pboyer 

### FYIs

(FILL ME IN, Optional) Names of anyone else you wish to be notified of

### CODE REVIEW

**Comment by pboyer** [comment]
Is this ready for review? It has a "WIP" tag.

**Comment by aparajit-pratap** [comment]
@pboyer I'm currently addressing merge conflicts with this PR and `master` branch right now. I guess I'll ping you once these are resolved. Thanks for checking in.

File: src/Engine/ProtoImperative/Properties/Resources.resx [hunk 1]
```diff
@@ -132,6 +132,9 @@
   <data name="FunctionPointerNotAllowedAtBinaryExpression" xml:space="preserve">
     <value>Function pointer is not allowed at binary expression other than assignment!</value>
   </data>
+  <data name="ImperativeSymbolsAreReadOnly" xml:space="preserve">
+    <value>Variables declared in outer scope are read-only inside imperative block!</value>
+  </data>
```
Comment (pboyer) [inline src/Engine/ProtoImperative/Properties/Resources.resx]: I don't mean to bikeshed it, but I think this is a bit more clear. You might ask @kronz or @Racel for thoughts:

`Variables declared in an outer scope cannot be changed inside an imperative block.`

File: test/Engine/ProtoTest/ProtoAST/RoundTripTests.cs [hunk 2]
```diff
@@ -295,19 +295,19 @@ public void TestAstToCode()
...
-                "x = Point.ByCoordinates({1,2,3}, {4,5,6}<1>, {7, 8, 9}<2>)<0>;",
+                "x = Point.ByCoordinates([1,2,3], [4,5,6]<1>, [7, 8, 9]<2>)<0>;",
                 "x = this.foo(1, 2);",
                 "x = this.X;",
-                "return = {1, 2, 3};",
+                "return = [1, 2, 3];",
                 "return = Math.PI;",
                 "return = Math.Cos(1.2);", 
-                "x[0][1] = {};",
+                "x[0][1] = [];",
```
Comment (pboyer) [inline test/Engine/ProtoTest/ProtoAST/RoundTripTests.cs]: I'm surprised this wasn't an error I encountered before... Did you just revive this test?

File: src/Engine/ProtoImperative/CodeGen.cs [hunk 3]
```diff
@@ -1279,8 +1298,84 @@ traverse E
...
+            {
+                MakeSymbolsLocalInImperativeBlock(elseIfBlock.Body);
+            }
+        }
+
+        /// <summary>
+        /// Identify lhs symbols used in current codeblock that are allocated in 
+        /// a parent associative block and allocate a copy of them in the outer imperative block
+        /// </summary>
+        /// <param name="body"></param>
```
Comment (pboyer) [inline src/Engine/ProtoImperative/CodeGen.cs]: If you have nothing to say about the param, just omit this.

File: src/Engine/ProtoImperative/Properties/Resources.resx (hunk 1 above)
Comment (aparajit-pratap) [reply src/Engine/ProtoImperative/Properties/Resources.resx]: Yes, the word `read-only` may be too technical for our audience. `bikeshed` is definitely for me :)

File: test/Engine/ProtoTest/ProtoAST/RoundTripTests.cs (hunk 2 above)
Comment (aparajit-pratap) [reply test/Engine/ProtoTest/ProtoAST/RoundTripTests.cs]: Oh I just manually updated it even though it's marked failure. I think your script skipped those tests that were marked as "failure"?

**Pull request merged**